pkg/install/error_test.go

```diff
@@ -0,0 +1,193 @@
+								"message": tt.message,
+							},
+						},
+					}, "", "", nil, "",
+				),
+			)
+			if isQuota != tt.wantBool || errMsg != tt.wantMessage {
+				t.Error(tt.wantBool, tt.wantMessage)
+			}
+		})
+	}
+}
+
+func TestIsDeploymentActiveError(t *testing.T) {
+	for _, tt := range []struct {
+		name          string
+		nilServiceErr bool
+		code          string
+		wantBool      bool
+	}{
+		{
+			name:          "nil ServiceError",
+			nilServiceErr: true,
+		},
+		{
+			name: "Different code in ServiceError",
+			code: "AnotherCode",
+		},
+		{
+			name:     "Deployment active",
+			code:     "DeploymentActive",
+			wantBool: true,
+		},
+	} {
+		t.Run(tt.name, func(t *testing.T) {
+			var isActive bool
+			if tt.nilServiceErr {
+				isActive = isDeploymentActiveError(
+					autorest.NewErrorWithError(azure.RequestError{}, "", "", nil, ""),
+				)
+			} else {
+				isActive = isDeploymentActiveError(
+					autorest.NewErrorWithError(
+						azure.RequestError{
+							ServiceError: &azure.ServiceError{
+								Code: tt.code,
+							},
+						}, "", "", nil, "",
+					),
+				)
+			}
+			if isActive != tt.wantBool {
+				t.Error(tt.wantBool)
+			}
+		})
+	}
+}
+
+func TestCheckDeploymentErr(t *testing.T) {
+	ctx := context.Background()
+	resourceGroup := "test"
+	deploymentName := "azuredeploy"
+	fakeQuotaExceededMessage := "Quota exceeded"
+	quotaErr := autorest.NewErrorWithError(
+		&azure.ServiceError{
+			Details: []map[string]interface{}{
+				{
+					"code":    "QuotaExceeded",
+					"message": fakeQuotaExceededMessage,
+				},
+			},
+		}, "", "", nil, "",
+	)
+	activeErr := autorest.NewErrorWithError(
+		azure.RequestError{
+			ServiceError: &azure.ServiceError{
+				Code: "DeploymentActive",
+			},
+		}, "", "", nil, "",
+	)
+	otherErr := errors.New("☹")
+
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+	m := mock_resources.NewMockDeploymentsClient(ctrl)
+
+	for _, tt := range []struct {
+		name          string
+		originalErr   error
+		activeWaitErr error
+		wantErr       error
+	}{
+		{
+			name:        "Deployment active, then wait successful",
+			originalErr: activeErr,
+		},
+		{
+			name:          "Deployment active, then timeout",
+			originalErr:   activeErr,
+			activeWaitErr: wait.ErrWaitTimeout,
+			wantErr:       wait.ErrWaitTimeout,
+		},
+		{
+			name:        "Resource quota exceeded",
+			originalErr: quotaErr,
+			wantErr:     api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeQuotaExceeded, fakeQuotaExceededMessage, ""),
+		},
+		{
+			name:          "Deployment active, then resource quota exceeded",
+			originalErr:   activeErr,
+			activeWaitErr: quotaErr,
+			wantErr:       api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeQuotaExceeded, fakeQuotaExceededMessage, ""),
+		},
+		{
+			name:        "Other error",
+			originalErr: otherErr,
+			wantErr:     otherErr,
+		},
+	} {
+		t.Run(tt.name, func(t *testing.T) {
+			if reflect.DeepEqual(tt.originalErr, activeErr) {
```

Comment:
We use this pattern a lot in other places:

```go
controller := gomock.NewController(t)
defer controller.Finish()

deploymentsMock := mock_resources.NewMockDeploymentsClient(controller)

if tt.mocks != nil {
	tt.mocks(deploymentsMock)
}
```

And if in the case you expect a call, you define the `mocks` function:

```go
{
	name:        "Deployment active, then wait successful",
	originalErr: activeErr,
	mocks:       func(deploymentsClient mock_resources.MockDeploymentsClient) {
		deploymentsClient.EXPECT().
			Wait(ctx, resourceGroup, deploymentName).
			Return(nil)
	}
},
```

Makes it easy to understand what we expect in each specific case.

Reply:
Yeah, much easier to read that way. Fixed.